src/index.ts

```diff
@@ -2451,6 +2478,113 @@ async function makeReclaimGasless({
 	}
 }
 
+/**
+ * gets the contract type
+ */
+async function getContractType({
+	provider,
+	address,
+}: {
+	provider: ethers.providers.Provider
+	address: string
+}): Promise<interfaces.EPeanutLinkType> {
+	const minimalABI = [
+		'function supportsInterface(bytes4) view returns (bool)',
+		'function totalSupply() view returns (uint256)',
+		'function balanceOf(address) view returns (uint256)',
+	]
+
+	// Interface Ids for ERC721 and ERC1155
+	const ERC721_INTERFACE_ID = '0x80ac58cd' // ERC721
+	const ERC1155_INTERFACE_ID = '0xd9b67a26' // ERC1155
+
+	const contract = new ethers.Contract(address, minimalABI, provider)
+
+	const isERC721 = await supportsInterface(contract, ERC721_INTERFACE_ID)
+	const isERC1155 = await supportsInterface(contract, ERC1155_INTERFACE_ID)
+	let isERC20 = false
+
+	// Check for ERC20 if it's not ERC721 or ERC1155
+	if (!isERC721 && !isERC1155) {
+		isERC20 = await contract
+			.totalSupply()
+			.then(() => true)
+			.catch(() => false)
+	}
+
+	if (address.toLowerCase() === ethers.constants.AddressZero.toLowerCase()) return 0
+	if (isERC20) return 1
+	if (isERC721) return 2
+	if (isERC1155) return 3
+}
+
+async function supportsInterface(contract, interfaceId) {
+	try {
+		return await contract.supportsInterface(interfaceId)
+	} catch (error) {
+		return false
+	}
+}
+
+async function getContractDetails({
+	address,
+	provider,
+}: {
+	address: string
+	provider: ethers.providers.Provider
+}): Promise<{ type: interfaces.EPeanutLinkType; decimals?: number; name?: string; symbol?: string }> {
+	//get the contract type
+	const contractType = await getContractType({ address: address, provider: provider })
+	//@ts-ignore
+	const batchprov = new ethers.providers.JsonRpcBatchProvider(provider.connection.url)
+
+	config.verbose && console.log('contractType: ', contractType)
+	switch (contractType) {
+		case 0: {
+			return {
+				type: 0,
+				decimals: 18,
+			}
+		}
+		case 1: {
+			const contract = new ethers.Contract(address, ERC20_ABI, batchprov)
+			const [name, symbol, decimals] = await Promise.all([
```

Comment:
These get batched with Batch Provider? Is that how it works? Aren't they distinct promised? I struggle to get how this works, but if it's batched, then that's great

Reply:
We looked into this back in lisbon, yes it gets batched. Not sure how, but it does xd